### PR TITLE
Fine tuning the docker environment.

### PR DESCRIPTION
**Pull request by philipsoutham**
- Locking down the containers by restricting all and explicitly allowing kernel capabilities.
- Making containers `read_only` when base image allows (some poorly built ones do not)
  - Also explicitly defining volumes where data should be stored/mutated.
- Making sure uid is not root when we can
- Defining configuration values in environment variables
- Taking the stuff from `tiangolo/uvicorn-gunicorn-fastapi:python3.7` that makes it special and building our own image from the official python base image
- using the official redis base image

### CODE REVIEW

**Comment by vlall** [comment]
Getting this error:
```
[2021-04-23 05:46:39 +0000] [1] [INFO] Starting gunicorn 20.0.4
[2021-04-23 05:46:39 +0000] [1] [ERROR] Retrying in 1 second.
[2021-04-23 05:46:40 +0000] [1] [ERROR] Retrying in 1 second.
[2021-04-23 05:46:41 +0000] [1] [ERROR] Retrying in 1 second.
[2021-04-23 05:46:42 +0000] [1] [ERROR] Retrying in 1 second.
[2021-04-23 05:46:43 +0000] [1] [ERROR] Retrying in 1 second.
[2021-04-23 05:46:44 +0000] [1] [ERROR] Can't connect to ('0.0.0.0', 80)
{"loglevel": "info", "workers": 2, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}
```

**Comment by vlall** [comment]
```
PermissionError: [Errno 13] Permission denied: '/app/log/app.log'
[2021-04-23 05:55:20 +0000] [10] [INFO] Worker exiting (pid: 10)
[2021-04-23 05:55:20 +0000] [38] [INFO] Worker exiting (pid: 38)
{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}
{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 209, in run
    self.sleep()
{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 357, in sleep
    ready = select.select([self.PIPE[0]], [], [], 1.0)
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/gunicorn", line 8, in <module>
    sys.exit(run())
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/wsgiapp.py", line 58, in run
    WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/base.py", line 228, in run
    super().run()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/app/base.py", line 72, in run
    Arbiter(self).run()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 229, in run
    self.halt(reason=inst.reason, exit_status=inst.exit_status)
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 342, in halt
    self.stop()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 393, in stop
    time.sleep(0.1)
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
```

**Comment by philipsoutham** [comment]
I'll see if I can reproduce

-------- Original Message --------

**Comment by philipsoutham** [comment]
Are you sure you're on the right branch? Have you rebuilt the container (`docker-compose up --build`)? I cannot duplicate

```shell
$ docker-compose logs -f fastapi
Attaching to fastapi
fastapi          | Checking for script in /app/prestart.sh
fastapi          | There is no script /app/prestart.sh
fastapi          | [2021-04-26 17:39:37 +0000] [1] [INFO] Starting gunicorn 20.1.0
fastapi          | [2021-04-26 17:39:37 +0000] [1] [INFO] Listening at: http://0.0.0.0:8080 (1)
fastapi          | [2021-04-26 17:39:37 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
fastapi          | [2021-04-26 17:39:37 +0000] [8] [INFO] Booting worker with pid: 8
fastapi          | [2021-04-26 17:39:37 +0000] [9] [INFO] Booting worker with pid: 9
fastapi          | [2021-04-26 17:39:37 +0000] [33] [INFO] Booting worker with pid: 33
fastapi          | [2021-04-26 17:39:37 +0000] [21] [INFO] Booting worker with pid: 21
fastapi          | [2021-04-26 17:39:37 +0000] [34] [INFO] Booting worker with pid: 34
fastapi          | [2021-04-26 17:39:37 +0000] [35] [INFO] Booting worker with pid: 35
fastapi          | [2021-04-26 17:39:37 +0000] [36] [INFO] Booting worker with pid: 36
fastapi          | [2021-04-26 17:39:37 +0000] [61] [INFO] Booting worker with pid: 61
fastapi          | [2021-04-26 17:39:37 +0000] [71] [INFO] Booting worker with pid: 71
fastapi          | [2021-04-26 17:39:37 +0000] [83] [INFO] Booting worker with pid: 83
fastapi          | [2021-04-26 17:39:38 +0000] [95] [INFO] Booting worker with pid: 95
fastapi          | [2021-04-26 17:39:38 +0000] [107] [INFO] Booting worker with pid: 107
fastapi          | [2021-04-26 17:39:41 +0000] [9] [INFO] Started server process [9]
fastapi          | [2021-04-26 17:39:41 +0000] [9] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:41 +0000] [9] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:41 +0000] [35] [INFO] Started server process [35]
fastapi          | [2021-04-26 17:39:41 +0000] [35] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:41 +0000] [35] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [33] [INFO] Started server process [33]
fastapi          | [2021-04-26 17:39:42 +0000] [33] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [33] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [8] [INFO] Started server process [8]
fastapi          | [2021-04-26 17:39:42 +0000] [8] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [8] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [95] [INFO] Started server process [95]
fastapi          | [2021-04-26 17:39:42 +0000] [95] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [83] [INFO] Started server process [83]
fastapi          | [2021-04-26 17:39:42 +0000] [107] [INFO] Started server process [107]
fastapi          | [2021-04-26 17:39:42 +0000] [107] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [83] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [95] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [83] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [107] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [21] [INFO] Started server process [21]
fastapi          | [2021-04-26 17:39:42 +0000] [21] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [21] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [34] [INFO] Started server process [34]
fastapi          | [2021-04-26 17:39:42 +0000] [34] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [34] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [71] [INFO] Started server process [71]
fastapi          | [2021-04-26 17:39:42 +0000] [71] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [71] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [36] [INFO] Started server process [36]
fastapi          | [2021-04-26 17:39:42 +0000] [36] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [36] [INFO] Application startup complete.
fastapi          | [2021-04-26 17:39:42 +0000] [61] [INFO] Started server process [61]
fastapi          | [2021-04-26 17:39:42 +0000] [61] [INFO] Waiting for application startup.
fastapi          | [2021-04-26 17:39:42 +0000] [61] [INFO] Application startup complete.
```

File: docker-compose.yml [hunk 1]
```diff
@@ -1,21 +1,101 @@
...
+      - ALL
     ports:
-      - "8080:80"
+      - "8080:8080"
     networks:
       - "jointops"
     volumes:
-      - type: bind
-        source: ./api/data
-        target: /app/data
```
Comment (vlall) [inline docker-compose.yml:15]: I need to keep the bind to `./api/data`, can i just re-add this portion?
```
      - type: bind
         source: ./api/data
```

Comment (philipsoutham) [reply docker-compose.yml:15]: you can just add

```
      - "./api/data:/api/data"
```

under `volumes:`, you don't need to use the long form of declaring it.

**Pull request closed without merging**